### PR TITLE
Support the 'time_major' param for the new cuDNN RNN API

### PR DESCRIPTION
**Pull request by kaixih**
We added the `time_major` param for the new cuDNN RNN API (based on the PR https://github.com/tensorflow/tensorflow/pull/23588). As in `tf.nn.dynamic_rnn()`, the `time_major` controls input format to [max_time, batch_size, num_inputs] if it is `True` or [batch_size, max_time, num_inputs] if `False`. The `time_major` will only used when `sequence_lengths` is provided (meaning the new cuDNN RNN API is used). This is because only the new cuDNN RNN API supports both time major and batch major formats (https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnSetRNNDataDescriptor).

### CODE REVIEW

**Comment by houtoms** [comment]
Any updates on the PR?

**Comment by nluehr** [comment]
Attention @azaks2 
Can we get this PR reviewed?

**Comment by tfboyd** [comment]
@azaks2 will find someone.

**Comment by tfboyd** [comment]
@protoget ping.

**Comment by protoget** [comment]
Thanks for the PR. 

1. For `dynamic_rnn`, the `time_major` option does not affect any model weights and there is not checkpoint compatibility issue. Does switching between time_major and data_major affect Cudnn opaque param layout?
 
2. We should not have `time_major` depending on sequence_length. V3 kernel is  called if seq_len is not None, or time_major is False.

File: tensorflow/contrib/cudnn_rnn/python/layers/cudnn_rnn.py
Comment (protoget) [inline tensorflow/contrib/cudnn_rnn/python/layers/cudnn_rnn.py]: Any unittests at layer level?

File: tensorflow/contrib/cudnn_rnn/python/kernel_tests/cudnn_rnn_ops_test.py
Comment (protoget) [inline tensorflow/contrib/cudnn_rnn/python/kernel_tests/cudnn_rnn_ops_test.py]:
```py
shape = [time, batch, input_size] if time_major else [batch, time, inpput_size]
```

save the switches.

File: tensorflow/contrib/cudnn_rnn/python/kernel_tests/cudnn_rnn_ops_test.py
Comment (protoget) [inline tensorflow/contrib/cudnn_rnn/python/kernel_tests/cudnn_rnn_ops_test.py]: axis = 0 if time_major else 1

Comment (protoget) [reply tensorflow/contrib/cudnn_rnn/python/kernel_tests/cudnn_rnn_ops_test.py]: similar for other places.

File: tensorflow/contrib/cudnn_rnn/python/kernel_tests/cudnn_rnn_ops_test.py
Comment (protoget) [reply tensorflow/contrib/cudnn_rnn/python/kernel_tests/cudnn_rnn_ops_test.py]: similar for other places.

**Review by protoget — COMMENTED** [review]
Added a few comments.

**Comment by ebrevdo** [comment]
Most TF rnn ops are batch major.  Perhaps time_major should be false by default?

**Comment by ebrevdo** [comment]
(... are batch major by default)

**Comment by houtoms** [comment]
Hi @protoget , updated the codes to solve all your suggestions.
Below are my answers to your questions:
* For `dynamic_rnn`, the `time_major` option does not affect any model weights and there is not checkpoint compatibility issue. Does switching between time_major and data_major affect Cudnn opaque param layout?
It will not change the layout.
* We should not have `time_major` depending on sequence_length. V3 kernel is  called if seq_len is not None, or time_major is False.
I think time_major has to depend on sequence_length for now. For example, if users set time_major as False and provide no sequence_lengths, we still cannot call v3 kernel, because it requires sequence_lengths.
* Which major do you recommend?
I think by default, the time_major should be True, which requires no sequence_lengths parameter.

**Comment by protoget** [comment]
Re ebrevdo@, Cudnn used to only support time_major, so set default to time_major is for backwards compatibility. But that's fine since our keras UnifiedLSTM support both and we just need to whitelist CuDNN + time_major in the dispatch code.

**Comment by protoget** [comment]
Re houtoms@, thanks for responding.
2) When time_major is false, and no sequence_lengths are provided, then we should either consider all input examples are of the same length, and use v3.
3) Performance-wise, do you recommend time_major or batch_major?

**Comment by houtoms** [comment]
> Re houtoms@, thanks for responding.
> 2) When time_major is false, and no sequence_lengths are provided, then we should either consider all input examples are of the same length, and use v3.
> 3) Performance-wise, do you recommend time_major or batch_major?

(2) Sure. In this case, we will prepare the sequence_lengths parameter for users. Let me do another commit.
(3) The perf is more or less same for the two formats, even if we have internal transformation for the time_major data.

**Comment by houtoms** [comment]
Fixed an issue based on your suggestion (2) @protoget

**Comment by houtoms** [comment]
Hi @protoget, I realized the PR has some conflict with master. So, I just solved that conflict and made a new commit. Please check. Thanks.

**Comment by tfboyd** [comment]
@protoget ,  I realize the last update was 12 hours ago.   Pinging because it is meet with NVIDIA day. :-)  It looks like we are really close.

File: tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py
Comment (protoget) [inline tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py]: This only works if the inputs is of static shape (shape known at graph-construction time).

To make it work for a dynamic shaped inputs (and please test w/ unittest. You can try a `tf.placeholder(shape=[None, None, None])` and run a session with feed_dict), try something like the following:
```py
batch_size, max_len = tf.get_shape(inputs)[0], tf.get_shape(inputs)[1]
seq_lengths = tf.fill([batch_size], max_len)
```

File: tensorflow/core/framework/op_kernel.cc
Comment (protoget) [inline tensorflow/core/framework/op_kernel.cc]: How about just
```c++
return result != input_name_map_.end()
```

File: tensorflow/core/framework/op_kernel.h
Comment (protoget) [inline tensorflow/core/framework/op_kernel.h]: `there exists an input`

File: tensorflow/core/kernels/cudnn_rnn_ops.cc
Comment (protoget) [inline tensorflow/core/kernels/cudnn_rnn_ops.cc]: Could you put all `const Tensor**` types close to each other?

File: tensorflow/contrib/cudnn_rnn/python/layers/cudnn_rnn.py
Comment (protoget) [reply tensorflow/contrib/cudnn_rnn/python/layers/cudnn_rnn.py]: ?

**Review by protoget — COMMENTED** [review]
LG in general. A few other comments. Should be good once they're addressed and passed tests.

**Comment by houtoms** [comment]
Solved the above comments. Please check. Thanks.

File: tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py
Comment (protoget) [inline tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py]: Could you add relevant tests?

**Review by protoget — COMMENTED** [review]
Almost done

File: tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py
Comment (houtoms) [reply tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py]: I think I have already added tests for this feature (with time_major parameter). My original commit has (in cudnn_rnn_ops_test.py), which force uses to provide both variable_seq_lengths and time_major.
```
if not variable_seq_lengths and not time_major:	
      self.skipTest("Batch major not supported")
```
But now I remove these lines (see commit 9c27326b98f975dd04908defd9c6925e9263aa78). So it has the tests on this branch (time_major but no variable_seq_length)

Comment (protoget) [reply tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py]: The tests didn't not cover dynamic shape input, thats why the previous commit's problem wasn't caught.

Comment (houtoms) [reply tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py]: OIC. Could you please point me to some example code in other ops unittest about adding the tf.placeholder and feed_dict?

Comment (houtoms) [reply tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py]: No worries. I found some examples. Will update it.

**Comment by houtoms** [comment]
Added the test cases to cover the dynamic shape input.

**Comment by houtoms** [comment]
Solved a new conflicts.

File: tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py
Comment (protoget) [inline tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py]: remove?

File: tensorflow/contrib/cudnn_rnn/python/kernel_tests/cudnn_rnn_ops_test.py
Comment (protoget) [inline tensorflow/contrib/cudnn_rnn/python/kernel_tests/cudnn_rnn_ops_test.py]: nitpick: it's a minor suggestion, up to you.
No need to create both nodes in the graph
```py
inputs_np = ...
if dynamic_shape_input:
  inputs = get_variable(...)
else:
  inputs = placeholder(...)
```

In fact, I think maybe placeholder alone is good -- if it works for dynamic shape input, it must work for static shape.

**Review by protoget — DISMISSED** [review]
LGTM! Thanks for all the hardwork.

File: tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py
Comment (houtoms) [reply tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py]: I merged from a previous commit bd36b48c555b2d46c41a179ed9f27a04806e9e66 for this change to solve the conflict.

File: tensorflow/contrib/cudnn_rnn/python/kernel_tests/cudnn_rnn_ops_test.py
Comment (houtoms) [reply tensorflow/contrib/cudnn_rnn/python/kernel_tests/cudnn_rnn_ops_test.py]: I thought this previously. But actually, I think we need to keep both nodes. I tried the dynamic inputs for rnn.dynamic_rnn() but it doesn't work. (I guess they cannot handle the loops dynamically?).

I can remove the static input tests if that is OK for you.

**Comment by houtoms** [comment]
Change some format issues based on the `Ubuntu Sanity` check results.

**Comment by houtoms** [comment]
@protoget I changed some python format issues yesterday. Could you take a look at the changes? Thanks.

**Review by protoget — DISMISSED** [review]
LGTM

**Comment by houtoms** [comment]
@protoget Found two indent format issues. Fixed. Thanks.

**Review by protoget — DISMISSED** [review]
LGTM

**Comment by houtoms** [comment]
Hi, @protoget It seems the merging is still blocked. Should we need someone to tag this as "kokoro:force-run"?

**Comment by tfboyd** [comment]
@rthadur Can you give us a status of the merge?  We have had some trouble lately and it would help to know if this is on track or if it is stuck.

**Comment by rthadur** [comment]
> @rthadur Can you give us a status of the merge? We have had some trouble lately and it would help to know if this is on track or if it is stuck.

@tfboyd still waiting for getting it merged internally , will take a look at it

**Comment by tfboyd** [comment]
@rthadur  Thank you kind person.  I do not deal with this very often so it is a black box where I know just enough to know how much I don't know about the flow.

**Comment by houtoms** [comment]
Thanks for your response.

**Comment by houtoms** [comment]
There are five tests failed and the logs says it is the "tests:api_compatibility_test". I think it is as expected because I did change the API?

**Comment by protoget** [comment]
You probably want to check this:
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/api/tests

**Comment by houtoms** [comment]
@protoget, Thanks. 

I have a question about this. From the logs, it seems the fails are because I added `time_major` for CudnnRNNV3 and CudnnRNNBackpropV3 operations, which I think is necessary. Do I need to avoid this?

**Comment by protoget** [comment]
Technically we should not alter the op signature. However I think V3 is not being used by anyone so I personally think it's ok. Otherwise you'd need to define a V4.

**Comment by houtoms** [comment]
Thanks. So, should we continue the merge?

**Comment by azaks2** [comment]
there are two questions:
1. backward compatability. You might be ok if none is really using the old version
2. API changes: you do need to update api def files, see protoget comment. Also you are change OpKernel api so you def. need an API owner approval. adding alextp.

**Comment by houtoms** [comment]
Do you mean the `tensorflow/core/api_def/base_api/api_def_CudnnRNNV3.pbtxt` and ` tensorflow/core/api_def/base_api/api_def_CudnnRNNBackpropV3.pbtxt`?
If yes, should I wait for @alextp 's response or can I make a new commit now?

File: tensorflow/core/framework/op_kernel.h
Comment (alextp) [inline tensorflow/core/framework/op_kernel.h]: Revert the changes to these files?

Comment (houtoms) [reply tensorflow/core/framework/op_kernel.h]: I got the suggestion from @protoget to add this checking here. Which one should I follow?

**Comment by houtoms** [comment]
Just updated the api_def files for the cuDNNRNN v3 and cuDNNRNNBackprop v3 by adding the time_major parameter.

File: tensorflow/core/framework/op_kernel.h
Comment (alextp) [reply tensorflow/core/framework/op_kernel.h]: I'm a lot more comfortable approving changes which do not touch core tf functionality like op_kernel.{h,cc}.

Since a quick search over the non-contrib part of your pull request didn't find any usages of these new symbols I think it's more prudent to drop them.

Comment (houtoms) [reply tensorflow/core/framework/op_kernel.h]: Sure. I reverted this changes on op_kernel in the new commit.

**Comment by houtoms** [comment]
Reverted the changes on op_kernel.h/cc based on @alextp suggestion.

**Comment by houtoms** [comment]
@alextp Any updates?

**Review by alextp — DISMISSED** [review]
Approving for API owners

**Comment by houtoms** [comment]
Thanks. Do we need someone to label it to start the automatic tests?

**Comment by houtoms** [comment]
I think the only failed tests are the API compatibility tests and I think these API changes have already been approved. Do I need to do anything else for this?

**Comment by houtoms** [comment]
@alexp Can you point me how to avoid the failure of the `//tensorflow/tools/api/tests:api_compatibility_test`, shown in the the Ubuntu Python 2/3?

**Comment by alextp** [comment]
@houtoms the test log says

```
This test ensures all changes to the public API of TensorFlow are intended.
If this test fails, it means a change has been made to the public API. Backwards
incompatible changes are not allowed. You can run the test as follows to update
test goldens and package them with your change.
    $ bazel build tensorflow/tools/api/tests:api_compatibility_test
    $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
You will need an API approval to make changes to the public TensorFlow API. This
includes additions to the API.
```

**Comment by houtoms** [comment]
Hi @alextp , Thanks very much for your clarification. I updated the goldens and packaged these changes.

**Review by alextp — APPROVED** [review]

**Comment by houtoms** [comment]
@alextp I check the merging is blocked by `import/copybara`. Is this the API change approval?

**Pull request merged**
